File: code/ch12/12.2.1.predictTheWinner.py
from typing import List


class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        sum = 0
        for i in nums:
            sum += i
        mem = []
        for i in range(len(nums)):
            mem.append([0] * len(nums))

        return 2 * self.find_max(0, len(nums) - 1, nums, mem) >= sum

    def find_max(
        self, left: int, right: int, nums: List[int], mem: List[List[int]]
    ) -> int:
        if left < 0 or right < 0 or left > right:
            return 0
        if mem[left][right] != 0:
            return mem[left][right]
        if left == right:
            mem[left][right] = nums[left]
            return nums[left]
        max_num = max(
            nums[left]
            + min(
                self.find_max(left + 2, right, nums, mem),
                self.find_max(left + 1, right - 1, nums, mem),
            ),
            nums[right]
            + min(
                self.find_max(left + 1, right - 1, nums, mem),
                self.find_max(left, right - 2, nums, mem),
            ),  # black formatter added trailing comma
        )
        mem[left][right] = max_num
        return max_num

File: code/ch12/12.2.1.PredictTheWinner.cpp
class Solution {
private:
    int find_max(int left, int right, vector<int>& nums, vector<vector<int>>& mem) {
        if (left < 0 || right < 0 || left > right)
            return 0;
        
        if (mem[left][right] != 0)
            return mem[left][right];
        
        if (left == right) {
            mem[left][right] = nums[left];
            return nums[left];
        }
        
        int max_num = max(nums[left]  + min(find_max(left + 2, right, nums, mem), find_max(left + 1, right - 1, nums, mem)),
                          nums[right] + min(find_max(left + 1, right - 1, nums, mem), find_max(left, right - 2, nums, mem)));
        mem[left][right] = max_num;
        
        return max_num;
    }
public:
    bool PredictTheWinner(vector<int>& nums) {
        int sum = 0;
        for (int i = 0; i < nums.size(); i++)
            sum += nums[i];
        
        vector<vector<int>> mem;
        for (int i = 0; i < nums.size(); i++)
            mem.push_back(vector<int>(nums.size(), 0));
        
        return 2 * find_max(0, nums.size() - 1, nums, mem) >= sum;
    }
};
